{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "edb50bd7-0a0d-4464-b495-cfd722ef8b6f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import inspect\n",
    "import seaborn\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import torch\n",
    "import scanpy as sc\n",
    "from tqdm import tqdm\n",
    "import sys\n",
    "import pickle\n",
    "import PyComplexHeatmap as pch\n",
    "import scvi\n",
    "import IPython\n",
    "import pandas as pd\n",
    "import scipy\n",
    "import numpy as np\n",
    "import itertools\n",
    "import xarray as xr\n",
    "sc.settings.figdir=os.path.expanduser('/allen/programs/celltypes/workgroups/rnaseqanalysis/HMBA/Aim1_Atlases/BasalGanglia_paper_package/analysis/dev_map')\n",
    "sc._settings.settings._vector_friendly=True\n",
    "\n",
    "device = torch.device(\"cuda:0\" if torch.cuda.is_available() else \"cpu\")\n",
    "\n",
    "if torch.cuda.is_available():\n",
    "    print(\"GPU is available\")\n",
    "    print(\"Number of GPUs:\", torch.cuda.device_count())\n",
    "    print(\"GPU Name:\", torch.cuda.get_device_name(0))\n",
    "else:\n",
    "    print(\"GPU is not available\")\n",
    "\n",
    "sys.path.append('/home/matthew.schmitz/utils/mts-utils/')\n",
    "from genomics import sc_analysis\n",
    "\n",
    "\n",
    "import antipode\n",
    "\n",
    "gs1=matplotlib.colors.ListedColormap(sc.pl.palettes.godsnot_102,name='godsnot_102')\n",
    "try:\n",
    "    matplotlib.colormaps.register(name='godsnot_102',cmap=gs1)\n",
    "except:\n",
    "    pass\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45a80129-bd5c-4d86-af27-53577cb34717",
   "metadata": {},
   "outputs": [],
   "source": [
    "table_dir = '/allen/programs/celltypes/workgroups/rnaseqanalysis/HMBA/Aim1_Atlases/BasalGanglia_paper_package/anno_tables/RNA'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "72caf54c-d004-4af3-8cc0-f620e093710c",
   "metadata": {},
   "outputs": [],
   "source": [
    "batch_key='load_name'\n",
    "species_key='organism'\n",
    "donor_key='donor_id'\n",
    "layer_key='UMIs'\n",
    "leaf_key='Group'\n",
    "MDE_KEY = \"X_umap_species_integrated\"\n",
    "sex_key = \"self_reported_sex\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "554cf389-045b-44ac-ad3d-05c59139995f",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = sc.read('/allen/programs/celltypes/workgroups/rnaseqanalysis/HMBA/Aim1_Atlases/BasalGanglia_paper_package/data/xspecies/Consensus_HMBA_basalganglia_AIT_pre-print.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b9cf64b2-a8fe-4b6c-a190-78d817e84a3d",
   "metadata": {},
   "outputs": [],
   "source": [
    "groups = pd.read_csv('/allen/programs/celltypes/workgroups/rnaseqanalysis/HMBA/Aim1_Atlases/BasalGanglia_paper_package/anno_tables/consensus_annotation_groups.tsv',header=None)[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a6b8cbc-eb42-4e80-a9ad-19d5de2415d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['Group'] = adata.obs['Group'].cat.reorder_categories(groups)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "da1d72ef-816d-4784-a144-6488be9efb0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs[sex_key] = adata.obs[sex_key].astype(str)\n",
    "adata.obs.loc[adata.obs[species_key]=='Marmoset',sex_key] = adata.obs.loc[adata.obs[species_key]=='Marmoset','donor_id']\n",
    "adata.obs.loc[adata.obs[species_key]=='Marmoset',sex_key]\n",
    "adata.obs['donor_id'].replace({'cjCroissant':'male','cjNutmeg':'male','cjJellybean':'female','cjRambo':'female'})\n",
    "adata.obs[sex_key] = adata.obs[sex_key].replace({'cjCroissant':'male','cjNutmeg':'male','cjJellybean':'female','cjRambo':'female'}).astype('string')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e9ca564-20eb-45b2-a185-836a78dc3a3c",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs[species_key] = adata.obs[species_key].astype('category')\n",
    "species_colors = {'Human':'#377eb8','Macaque':'#4daf4a','Marmoset':'#FF5F5D','Mouse':'#ffa300'}\n",
    "adata.uns[f'{species_key}_colors'] = [species_colors[x] for x in adata.obs[species_key].cat.categories]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ebf07843-6f71-4401-8b6a-9ae0e04626de",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['sex'] = adata.obs[sex_key].astype(str).str.lower().astype('category')\n",
    "sex_key = 'sex'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "55cf5151-24c6-4821-979e-663b7144b633",
   "metadata": {},
   "outputs": [],
   "source": [
    "colors = pd.read_csv('/allen/programs/celltypes/workgroups/rnaseqanalysis/HMBA/Aim1_Atlases/BasalGanglia_paper_package/anno_tables/Group_colors.txt',header=None,sep='\\t')\n",
    "adata.uns['Group_color_dict'] = dict(zip(colors[0],colors[1]))\n",
    "adata.uns['Group_colors'] = [adata.uns['Group_color_dict'].get(x,\"#777777\") for x in adata.obs['Group'].cat.categories]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9e679b41-8a56-414c-b4d3-da68d3511731",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.embedding(adata,color='Group',basis='X_umap')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "161692ac-fda2-4335-ac7e-ae9b7fc6d377",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.embedding(adata[adata.obs['Group'].str.contains('MSN|NUDAP',regex=True),:],color='Group',basis='X_umap',save='_MSNs_only.pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "94d713f4-73b2-4139-b152-e48c5721c1d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.embedding(adata[adata.obs['Group'].str.contains('MSN|NUDAP',regex=True),:],color=species_key,basis='X_umap',save='_species_MSNs_only.pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c3c1b3c0-2bba-4848-86fd-0651a61b78cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "if not os.path.exists(os.path.join(table_dir, \"species_group_means.netcdf\")):\n",
    "    # Get real means and xero proportions\n",
    "    log_real_means,real_mean_levels=antipode.model_functions.get_real_leaf_means(adata,species_key,leaf_key,layer=layer_key)\n",
    "    real_means=pd.DataFrame(log_real_means.mean(0),columns=adata.var.index,index=real_mean_levels[leaf_key])\n",
    "    real_means=real_means.loc[real_mean_levels[leaf_key],:]\n",
    "    aggr_zeros=antipode.model_functions.group_aggr_anndata(adata,[species_key,leaf_key],layer=layer_key,agg_func=antipode.model_functions.prop_zeros,normalize=True)\n",
    "    \n",
    "    coords = {species_key:list(adata.obs[species_key].cat.categories),'Group':list(adata.obs['Group'].cat.categories),'var':list(adata.var.index.astype(str))}\n",
    "    lrm = xr.DataArray(log_real_means,coords=coords)\n",
    "    az = xr.DataArray(aggr_zeros[0],coords=coords)\n",
    "    ads = xr.Dataset({'scalars':lrm,'proportions':az})\n",
    "    ads.to_netcdf(os.path.join(table_dir, \"species_group_means.netcdf\"), engine=\"scipy\")\n",
    "else:\n",
    "    ads = xr.open_dataset(os.path.join(table_dir, \"species_group_means.netcdf\"), engine=\"scipy\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e03c6f64-f74e-4d58-88af-c56d966afa6a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# hann mapping totally failed\n",
    "# import v1utils\n",
    "# mapping_df = v1utils.import_hann('/allen/programs/celltypes/workgroups/rnaseqanalysis/HMBA/Aim1_Atlases/BasalGanglia_paper_package/data/human/Human_HMBA_basalganglia_AIT_pre-print_WB_MAPPING/hann_results.json')\n",
    "# # mapping_df.index = list(adata.obs['cell_id'])\n",
    "# mapping_df = mapping_df.loc[~mapping_df.index.duplicated(),:]\n",
    "# # [\"class_label\",\"subclass_label\",\"cluster_label\"]\n",
    "# for x in mapping_df.columns:\n",
    "#     adata.obs[x]=mapping_df[x]\n",
    "#     adata.obs[x] = [str(i) for i in adata.obs[x]]\n",
    "#     adata.obs[x] = adata.obs[x].astype('category')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cabc60aa-28bf-460d-bcdd-72dd1383b1a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "dads = xr.open_dataset('/home/matthew.schmitz/Matthew/models/1.9.1.8.5_Dev_final_600clusters/species_class_means.netcdf', engine=\"scipy\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0b1c8e0d-bcd3-43b8-8e91-29c3efabe40c",
   "metadata": {},
   "outputs": [],
   "source": [
    "adult_means = np.exp(ads['scalars'].mean(species_key)-ads['scalars'].mean(species_key).max(leaf_key)).to_dataframe().unstack().droplevel(0, axis=1) #scaled linear-space pseudobulk\n",
    "dev_means = np.exp(dads['scalars'].mean('species')-dads['scalars'].mean('species').max('Initial_Class_markers_level_2')).to_dataframe().unstack().droplevel(0, axis=1) #scaled linear-space pseudobulk"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "46d83d09-9c40-4389-9e17-917db078b178",
   "metadata": {},
   "outputs": [],
   "source": [
    "tf_genes=pd.read_csv('/home/matthew.schmitz/Matthew/utils/zizhens_tf_code.txt',sep='\\t')\n",
    "include_genes=list(tf_genes['gene'].str.upper())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b881fdc7-beb7-4f6e-8de0-bfe3faca7acc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# gl_path='/allen/programs/celltypes/workgroups/rnaseqanalysis/EvoGen/Team/Matthew/data/taxtest/gene_lists/'\n",
    "# tf_genes=pd.read_csv(os.path.join(gl_path,'TFs_lambert_pmid29425488_1.01.txt'),sep='\\t')\n",
    "# include_genes=list(tf_genes['hgnc_symbol'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f9f00a9-c412-4958-b399-040c9dfd1d5b",
   "metadata": {},
   "outputs": [],
   "source": [
    "'' in include_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "09cbdc26-0ec4-44ad-ad27-cf02126f0108",
   "metadata": {},
   "outputs": [],
   "source": [
    "include_genes = include_genes + ['FOXP1','PBX3','TSHZ1','CASZ1','SOX1','SKOR1','SKOR2','DRD1','DRD2'] #genes missing from zizhen's list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "82ce1a19-ef35-4054-ad5f-3675849f1904",
   "metadata": {},
   "outputs": [],
   "source": [
    "include_genes = list(set(include_genes) & set(ads.coords['var'].data) & set(dads.coords['var'].data))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3ef96d03-edc5-4286-9f39-9a67bc244bca",
   "metadata": {},
   "outputs": [],
   "source": [
    "# include_genes = list(set(dev_means.loc[:,include_genes].std(0).sort_values().index[-500:]) & set(adult_means.loc[:,include_genes].std(0).sort_values().index[-500:]))\n",
    "# print(len(include_genes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9daa872f-4ba9-450c-b9cd-a4280ef3b67b",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(len(include_genes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "da566168-9d50-4573-9801-f3dfa869b089",
   "metadata": {},
   "outputs": [],
   "source": [
    "dev_means = dev_means.loc[~dev_means.index.str.contains('^Cb|^NPC|^Schwann|^L-Q|Progen_Astro|Progen_Tany|Progen_FP|Progen_Hem')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bc52595c-68e2-47b8-a493-a9c34e2560ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "corrs=sc_analysis.corr2_coeff(dev_means.loc[:,include_genes],adult_means.loc[:,include_genes])\n",
    "corrs=np.nan_to_num(corrs,0.)\n",
    "corrs=pd.DataFrame(corrs,index=dev_means.index,columns=adult_means.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ce613b00-d78e-4218-9b49-a981f6578a30",
   "metadata": {},
   "outputs": [],
   "source": [
    "seaborn.clustermap(adult_means.loc[:,adult_means.columns.isin(include_genes)],col_cluster=True,xticklabels=True,yticklabels=True)\n",
    "plt.show()\n",
    "seaborn.clustermap(dev_means.loc[:,dev_means.columns.isin(include_genes)],col_cluster=True)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9dcafd1d-88b6-447e-9797-5525351ce417",
   "metadata": {},
   "outputs": [],
   "source": [
    "g = seaborn.clustermap(corrs,cmap='coolwarm',figsize=(30,30),xticklabels=True,yticklabels=True,row_cluster=True,col_cluster=True)\n",
    "g.ax_heatmap.set_xticklabels(g.ax_heatmap.get_xmajorticklabels(), fontsize = 6)\n",
    "g.ax_heatmap.set_yticklabels(g.ax_heatmap.get_ymajorticklabels(), fontsize = 6)\n",
    "corrs.to_csv(os.path.join(sc.settings.figdir,'dev_corrs.csv'))\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "779dbe1a-1f1f-402e-9273-41f14cf57a53",
   "metadata": {},
   "outputs": [],
   "source": [
    "neuron_corrs = corrs.loc[~corrs.index.str.contains('Astro|Oligo|OPC|Mesench|Immune|CSF1|P2RY12|Ependy|Endo|Angio|SMC|Choroid|Peri|Hypendy|VLMC|ABC'),corrs.columns.str.contains('GABA|Glut|Dopa|Gly|MSN|Core|Shell')]\n",
    "g = seaborn.clustermap(neuron_corrs,cmap='coolwarm',figsize=(30,30),xticklabels=True,yticklabels=True,row_cluster=True,col_cluster=True)\n",
    "g.ax_heatmap.set_xticklabels(g.ax_heatmap.get_xmajorticklabels(), fontsize = 6)\n",
    "g.ax_heatmap.set_yticklabels(g.ax_heatmap.get_ymajorticklabels(), fontsize = 6)\n",
    "corrs.to_csv(os.path.join(sc.settings.figdir,'dev_neuron_corrs.csv'))\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "92a73bf5-3964-45cb-8d8f-46094674e842",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['Initial_Class_markers_level_2'] = adata.obs['Group'].replace(corrs.idxmax(0).to_dict())\n",
    "adata.obs['Initial_Class_markers_level_2'] = adata.obs['Initial_Class_markers_level_2'].str.replace('_[0-9]+$','',regex=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "75e645c4-5fa6-49e5-957a-52ffc63b3c61",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.embedding(adata,color='Initial_Class_markers_level_2',basis='X_umap',legend_loc='on data',legend_fontsize=5,save='initial_class.pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0abec04b-d44c-4481-8889-b739b305d452",
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.set_option('display.max_rows', 100)\n",
    "adata.obs.loc[(adata.obs['Neighborhood']!='Nonneuron')][['Group','Initial_Class_markers_level_2']].drop_duplicates().reset_index().drop('index',axis=1).to_csv(os.path.join(sc.settings.figdir,'mappings.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d89e8bd8-021d-4094-850a-a28e47d72515",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.embedding(adata,color=leaf_key,basis='X_umap',legend_loc='on data',legend_fontsize=4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ee0c9fe6-356f-4a41-a507-9789ec11b9e5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a6786905-125e-40dc-8c12-b771603c89cf",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5f2120bd-9f49-446a-84dd-80853df4cf03",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d19ffc81-3099-49d1-b9f0-f7c1f9a0a4b3",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "rapids-24.12",
   "language": "python",
   "name": "rapids-24.12"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
